File: LC 855.cpp
class ExamRoom {
    int n;
    vector<int> a;
public:
    ExamRoom(int n) {
        this->n = n;
    }
    
    int seat() {
        if (a.empty()) {
            a.push_back(0);
            return 0;
        }
        int maxDist = a[0];
        int seatToSit = 0;
        for (int i = 0; i + 1 < a.size(); i++) {
            int dist = (a[i+1] - a[i]) / 2;
            if (dist > maxDist) {
                maxDist = dist;
                seatToSit = a[i] + dist;
            }
        }
        // check the back 
        if (n - 1 - a.back() > maxDist) {
            seatToSit = n - 1;
        }
        auto it = lower_bound(a.begin(), a.end(), seatToSit);
        a.insert(it, seatToSit);
        return seatToSit;
    }
    
    void leave(int p) {
        auto it = find(a.begin(), a.end(), p);
        a.erase(it);
    }
};

/**
 * Your ExamRoom object will be instantiated and called as such:
 * ExamRoom* obj = new ExamRoom(n);
 * int param_1 = obj->seat();
 * obj->leave(p);
 */

File: 1000/LC 855.go
type ExamRoom struct {
    n int
    a []int
}


func Constructor(n int) ExamRoom {
    return ExamRoom{n: n}
}


func (this *ExamRoom) Seat() int {
    if len(this.a) == 0 {
        this.a = append(this.a, 0)
        return 0
    }
    maxDist := this.a[0]
    seatToSit := 0
    for i := 0; i + 1 < len(this.a); i++ {
        d := (this.a[i+1] - this.a[i]) / 2
        if d > maxDist {
            maxDist = d
            seatToSit = this.a[i] + d
        }
    }
    if this.n - 1 - this.a[len(this.a)-1] > maxDist {
        seatToSit = this.n - 1
    }
    this.a = append(this.a, seatToSit)
    sort.Ints(this.a)
    return seatToSit
}


func (this *ExamRoom) Leave(p int)  {
    for i, x := range this.a {
        if x == p {
            this.a = append(this.a[:i], this.a[i+1:]...)
            break
        }
    }
    return
}


/**
 * Your ExamRoom object will be instantiated and called as such:
 * obj := Constructor(n);
 * param_1 := obj.Seat();
 * obj.Leave(p);
 */